// @devexperts/rx-utils/dist/operators/filter-map.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var rxjs_1 = require("rxjs");
var Option_1 = require("fp-ts/lib/Option");
var FilterMapSubscriber = /** @class */ (function (_super) {
    tslib_1.__extends(FilterMapSubscriber, _super);
    function FilterMapSubscriber(destination, filterMap, thisArg) {
        var _this = _super.call(this, destination) || this;
        _this.filterMap = filterMap;
        _this.thisArg = thisArg;
        _this.count = 0;
        _this._next = function (value) {
            var result = undefined;
            try {
                result = _this.filterMap.call(_this.thisArg, value, _this.count++);
            }
            catch (e) {
                _this.destination.error && _this.destination.error(e);
                return;
            }
            if (result && Option_1.isSome(result)) {
                _this.destination.next && _this.destination.next(result.value);
            }
        };
        return _this;
    }
    return FilterMapSubscriber;
}(rxjs_1.Subscriber));
var FilterMapOperator = /** @class */ (function () {
    function FilterMapOperator(filterMap, thisArg) {
        this.filterMap = filterMap;
        this.thisArg = thisArg;
    }
    FilterMapOperator.prototype.call = function (subscriber, source) {
        return source.subscribe(new FilterMapSubscriber(subscriber, this.filterMap, this.thisArg));
    };
    return FilterMapOperator;
}());
/**
 * @deprecated Use `Filterable#filterMap`
 */
function filterMap(filterMap, thisArg) {
    return function mapOperation(source) {
        return source.lift(new FilterMapOperator(filterMap, thisArg));
    };
}
exports.filterMap = filterMap;
