// @devexperts/rx-utils/dist/sink2.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var rxjs_1 = require("rxjs");
var pipeable_1 = require("fp-ts/lib/pipeable");
var Apply_1 = require("fp-ts/lib/Apply");
var Array_1 = require("fp-ts/lib/Array");
exports.newSink = function (value, effects) { return ({ value: value, effects: effects }); };
exports.URI = '@devexperts/rx-utils//Sink';
exports.instanceSink = {
    URI: exports.URI,
    map: function (fa, f) { return exports.newSink(f(fa.value), fa.effects); },
    ap: function (fab, fa) { return exports.newSink(fab.value(fa.value), rxjs_1.merge(fab.effects, fa.effects)); },
    of: function (a) { return exports.newSink(a, rxjs_1.EMPTY); },
    chain: function (fa, f) {
        var fb = f(fa.value);
        return exports.newSink(fb.value, rxjs_1.merge(fa.effects, fb.effects));
    },
};
var getSemigroup = function (S) { return ({
    concat: function (x, y) { return exports.newSink(S.concat(x.value, y.value), rxjs_1.merge(x.effects, y.effects)); },
}); };
var getMonoid = function (M) { return (tslib_1.__assign(tslib_1.__assign({}, getSemigroup(M)), { empty: exports.instanceSink.of(M.empty) })); };
exports.sink = tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, exports.instanceSink), pipeable_1.pipeable(exports.instanceSink)), { sequenceT: Apply_1.sequenceT(exports.instanceSink), sequenceArray: Array_1.array.sequence(exports.instanceSink), getSemigroup: getSemigroup });
